### PR TITLE
experimental gitscanner api

### PR DESCRIPTION
**Pull request by technoweenie**
This PR builds on #1650 and introduces a `GitScanner` type. It's pretty empty, but it illustrates where I want to go with #1649. As of this PR, nothing outside of the `lfs` package uses `lfs.ScanRefs()` or `lfs.ScanRefsToChan()`.

TODO

* [x] merge #1650 
* [x] change base to `master`
* [x] :ship:

### CODE REVIEW

**Comment by ttaylorr** [comment]
This looks good to me.

**Comment by technoweenie** [comment]
I just pushed a change that stores the skipped refs on the `*GitScanner`. This means that initial git pushes with lots of refs will only calculate this data once, instead of on every `lfs.ScanRefs()` or `(*GitScanner) ScanLeftToRemote()` call.

**Comment by ttaylorr** [comment]
> I just pushed a change that stores the skipped refs on the `*GitScanner`.

Looks sweet. I'm a little bit worried about that list being potentially large, but not sure what other alternatives I'd go for.

**Comment by technoweenie** [comment]
Okay, I think this is ready to go now. `lfs/scanner.go` has no more public functions. Everything is using the new gitscanner api. I'd like to merge this so I can continue on more focused refactoring.

**Comment by technoweenie** [comment]
As a bonus, I exported the base channel wrapper stuff to `tools` (from #1625).

File: lfs/gitscanner.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	return &GitScanner{started: time.Now()}
+}
+
+// Close stops exits once all processing has stopped, and all resources are
+// tracked and cleaned up.
+func (s *GitScanner) Close() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.closed {
```
Comment (ttaylorr) [inline lfs/gitscanner.go]: I think you can remove the mutex lock and unlock here by using a `uint32` and either storing `0` or `1` by using the `sync/atomic` package.

File: lfs/gitscanner.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+func (s *GitScanner) ScanPreviousVersions(ref string, since time.Time) (*PointerChannelWrapper, error) {
+	return logPreviousSHAs(ref, since)
+}
+
+// ScanIndex scans the git index for modified LFS objects.
+func (s *GitScanner) ScanIndex(ref string) (*PointerChannelWrapper, error) {
+	return scanIndex(ref)
+}
+
+func (s *GitScanner) opts(mode ScanningMode) *ScanRefsOptions {
```
Comment (ttaylorr) [inline lfs/gitscanner.go]: Leaving a note here that this is a temporary measure at least until we can get rid of `*ScanRefsOptions`.

**Pull request merged**